rest/route/host.go

```diff
@@ -314,6 +316,134 @@ func getLimit(vals url.Values) (int, error) {
 	return limit, nil
 }
 
+////////////////////////////////////////////////////////////////////////
+//
+// POST /users/{user_id}/offboard_user
+
+func makeOffboardUser(sc data.Connector, env evergreen.Environment) gimlet.RouteHandler {
+	return &offboardUserHandler{
+		sc:  sc,
+		env: env,
+	}
+}
+
+type offboardUserHandler struct {
+	user   string
+	dryRun bool
+
+	env evergreen.Environment
+	sc  data.Connector
+}
+
+func (ch offboardUserHandler) Factory() gimlet.RouteHandler {
+	return &offboardUserHandler{
+		sc:  ch.sc,
+		env: ch.env,
+	}
+}
+
+func (ch *offboardUserHandler) Parse(ctx context.Context, r *http.Request) error {
+	ch.user = gimlet.GetVars(r)["user_id"]
+	if ch.user == "" {
+		return gimlet.ErrorResponse{
+			Message:    "user cannot be empty",
+			StatusCode: http.StatusBadRequest,
+		}
+	}
+	vals := r.URL.Query()
+	ch.dryRun = vals.Get("dry_run") == "true"
+
+	return nil
+}
+
+func (ch *offboardUserHandler) Run(ctx context.Context) gimlet.Responder {
+	usr := gimlet.GetUser(ctx)
+	opts := model.APIHostParams{
+		UserSpawned: true,
+	}
+	// returns all up-hosts for user
+	hosts, err := ch.sc.FindHostsInRange(opts, ch.user)
+	if err != nil {
+		return gimlet.NewJSONErrorResponse(errors.Wrap(err, "database error getting hosts"))
+	}
+
+	volumes, err := ch.sc.FindVolumesByUser(ch.user)
+	if err != nil {
+		return gimlet.NewJSONErrorResponse(errors.Wrap(err, "database error getting volumes"))
+	}
+
+	toTerminate := model.APIOffboardUserResults{
+		TerminatedHosts:   []string{},
+		TerminatedVolumes: []string{},
+	}
+	currentTime := time.Now()
+
+	mgrCache := map[cloud.ManagerOpts]cloud.Manager{}
+	for _, v := range volumes {
+		if !v.NoExpiration && v.Expiration.Before(currentTime) { // already terminated
+			continue
+		}
+		toTerminate.TerminatedVolumes = append(toTerminate.TerminatedVolumes, v.ID)
+		if ch.dryRun {
+			continue
+		}
+		mgrOpts := cloud.ManagerOpts{
+			Provider: evergreen.ProviderNameEc2OnDemand,
+			Region:   cloud.AztoRegion(v.AvailabilityZone),
+		}
+		mgr, ok := mgrCache[mgrOpts]
+		if !ok {
+			mgr, err = cloud.GetManager(ctx, ch.env, mgrOpts)
+			if err != nil {
+				return gimlet.MakeJSONInternalErrorResponder(err)
+			}
+			mgrCache[mgrOpts] = mgr
+		}
+
+		// detach volume first
+		if v.Host != "" {
+			for _, h := range hosts {
+				if h.Id == v.Host {
+					if err = mgr.DetachVolume(ctx, &h, v.ID); err != nil {
+						return gimlet.MakeJSONInternalErrorResponder(errors.Wrapf(err,
+							"error detaching volume '%s' from host '%s'", v.ID, h.Id))
+					}
+					break
+				}
+			}
+		}
+		if err = mgr.DeleteVolume(ctx, &v); err != nil {
+			return gimlet.MakeJSONInternalErrorResponder(err)
```

Comment:
maybe we could continue on error for `DeleteVolume` and `TerminateInstance`?

Reply:
Hmm I thought about that but I want to make sure stuff gets deleted, and that they retry in the case of error. Maybe if the error is NotFound then I'll just delete the host document, otherwise I'll still error so that the user knows.

Reply:
do we know for sure that MANA will retry if we return an error? (if not we would end up leaving over anything after this volume)

Reply:
🤷  it's up to them how they want to use it, but they should. The odds of this happening are very low anyway considering we filter out VolumeNotFound. How about we only return an error if it's for an unexpirable host/volume?

Reply:
and I can use a catcher